### PR TITLE
add mock server to integration test

### LINKED ISSUE #131: Add `integration test` to CI
# Context
Currently we have integration tests, but the test have not been added to CI yet(we can only run integration tests manually). I think it will be easy to add to CI once the mock server is ready.

- Start a mock server(like: [httpmock](https://github.com/alexliesenfeld/httpmock)) (for specified feature only)

### PR DESCRIPTION
**Pull request by zTgx**
- [x] add httpmock lib
- [x] starting mock server instance
- [x] refactor code structures
- [x] run all test cases

### CODE REVIEW

**Comment by Kailai-Wang** [comment]
Hi @zTgx, thanks for the PR, just a few small "style" things before we actually look into PR:
- we normally use [autoclose keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to link and close the issue
- if a PR is not yet ready, we either prepare it locally or convert it to "draft"
- I see there're fmt errors, you might want to consider installing the [githooks](https://github.com/litentry/tee-worker/blob/dev/.githooks/pre-commit) to help you formatting it automatically
- we normally have a feature branch and submit PR from that branch - but if you prefer the fork way, that's fine too

Hope that helps 🎉

**Comment by zTgx** [comment]
> 

really helpful, thanks for your advice. will fix it according to those conventions.

**Review by ziming-zung — CHANGES_REQUESTED** [review]
Some code needs to be modified.
- start mockserver in unit test and make the test pass

-  The [query_tweet ](https://github.com/litentry/tee-worker/blob/7f508be98bd574957ad0103544e964565bd47f40/litentry/core/data-providers/src/twitter_official.rs#L110-L118)API mock value  can refer to: https://github.com/litentry/tee-worker/blob/7f508be98bd574957ad0103544e964565bd47f40/litentry/core/identity-verification/src/lib.rs#L52-L57 Please note that `challenge_code` should be hardcoded when compiling with the `mockserver` feature. https://github.com/litentry/tee-worker/blob/7f508be98bd574957ad0103544e964565bd47f40/app-libs/stf/src/helpers.rs#L138-L140

File: litentry/core/data-providers/src/lib.rs [hunk 1]
```diff
@@ -64,13 +64,13 @@ pub mod base_url {
...
-// 	pub(crate) const TWITTER_LITENTRY: &str = "http://localhost";
-//
-// 	pub(crate) const DISCORD_OFFICIAL: &str = "http://localhost";
-// }
+#[cfg(test)]
+pub mod base_url {
+	pub(crate) const TWITTER_OFFICIAL: &str = "http://localhost:9527";
+	pub(crate) const TWITTER_LITENTRY: &str = "http://localhost:9527";
+	pub(crate) const DISCORD_OFFICIAL: &str = "http://localhost:9527";
+	pub(crate) const DISCORD_LITENTRY: &str = "http://localhost:9527";
```
Comment (ziming-zung) [inline litentry/core/data-providers/src/lib.rs:73]: I think this mod(base_url ) should work when compiled with mock-server feature

File: litentry/core/data-providers/src/twitter_litentry.rs [hunk 2]
```diff
@@ -46,7 +46,7 @@ impl Default for TwitterLitentryClient {
 
 #[derive(Serialize, Deserialize, Debug)]
 pub struct CheckFollow {
-	#[serde(rename(deserialize = "data"))]
+	// #[serde(rename(deserialize = "data"))]
```
Comment (ziming-zung) [inline litentry/core/data-providers/src/twitter_litentry.rs]: Why comment this marco(serde)

**Review by ziming-zung — CHANGES_REQUESTED** [review]
Basiclly, looks good except for a few small things. At last, I don't know how to start the mock-server. :sweat_smile:

File: litentry/core/data-providers/src/discord_official.rs [hunk 3]
```diff
@@ -103,5 +106,35 @@ mod tests {
 		// let message_id = "".as_bytes().to_vec();
 		// let result = client.query_message(channel_id, message_id);
 		// assert!(result.is_ok(), "query discord error: {:?}", result);
```
Comment (BillyWooo) [inline litentry/core/data-providers/src/discord_official.rs]: maybe we can remove this part

File: litentry/core/data-providers/src/discord_litentry.rs [hunk 4]
```diff
@@ -88,26 +86,52 @@ impl DiscordLitentryClient {
...
+		server.mock(|when, then| {
+			when.method(GET)
+				.path(path)
+				.query_param("guildid", "919848390156767232")
+				.query_param("handler", "againstwar#4779");
+			then.status(200).body(serde_json::to_string(&body).unwrap());
+		});
+
+		let guildid = "919848390156767232";
+		// let handler = "againstwar%234779";
```
Comment (BillyWooo) [inline litentry/core/data-providers/src/discord_litentry.rs]: maybe remove this line.

**Comment by Kailai-Wang** [comment]
The basic logic looks good, thanks 👍 

I'm a bit confused about the `mockserver` feature: it seems to be only used for generating a mock challenge code but you need to re-compile everything to make it work.

Depending on how we do the integration tests can we pass in an extra argument to tell if we are in mock mode or not, and then have different logic? Or even a different fn is more clear, something like `generate_challenge_code_mock`

And yes please remove the unneeded/commented code 🙏

File: litentry/core/data-providers/src/discord_litentry.rs (hunk 4 above)
Comment (deprecated1234) [reply litentry/core/data-providers/src/discord_litentry.rs]: we need to move this guildid into configs, and out of codes

**Review by ziming-zung — APPROVED** [review]
LGTM

**Comment by jingleizhang** [comment]
LGTM

**Pull request merged**